Userland/Libraries/LibWeb/HTML/Scripting/Fetching.cpp

```diff
@@ -38,43 +40,204 @@ String module_type_from_module_request(JS::ModuleRequest const& module_request)
+            auto scope_imports_match = TRY(resolve_imports_match(normalized_specifier, as_url, scope_imports));
+
+            // 2. If scopeImportsMatch is not null, then return scopeImportsMatch.
+            if (scope_imports_match.has_value())
+                return *scope_imports_match;
+        }
+    }
+
+    // 10. Let topLevelImportsMatch be the result of resolving an imports match given normalizedSpecifier, asURL, and importMap's imports.
+    auto top_level_imports_match = TRY(resolve_imports_match(normalized_specifier, as_url, import_map.imports()));
+
+    // 11. If topLevelImportsMatch is not null, then return topLevelImportsMatch.
+    if (top_level_imports_match.has_value())
+        return *top_level_imports_match;
+
+    // 12. If asURL is not null, then return asURL.
+    if (as_url.has_value())
+        return *as_url;
+
+    // 13. Throw a TypeError indicating that specifier was a bare specifier, but was not remapped to anything by importMap.
+    return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, String::formatted("Specifier '{}' is bare specifier but was not remapped by importMap.", specifier) };
+}
+
+// https://html.spec.whatwg.org/multipage/webappapis.html#resolving-an-imports-match
+WebIDL::ExceptionOr<Optional<AK::URL>> resolve_imports_match(String const& normalized_specifier, Optional<AK::URL> as_url, ModuleSpecifierMap const& specifier_map)
+{
+    // 1. For each specifierKey → resolutionResult of specifierMap:
+    for (auto const& entry : specifier_map) {
+        auto const& specifier_key = entry.key;
+        auto const& resolution_result = entry.value;
+
+        // 1. If specifierKey is normalizedSpecifier, then:
+        if (specifier_key == normalized_specifier) {
+            // 1. If resolutionResult is null, then throw a TypeError indicating that resolution of specifierKey was blocked by a null entry.
+            if (!resolution_result.has_value())
+                return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, String::formatted("Import resolution of '{}' was blocked by a null entry.", specifier_key) };
+
+            // 2. Assert: resolutionResult is a URL.
+            // FIXME: Clarify if "is a URL" means that the URL must be valid.
+            VERIFY(resolution_result->is_valid());
+
+            // 3. Return resolutionResult.
+            return *resolution_result;
+        }
+
+        // 2. If all of the following are true:
+        if (
+            // specifierKey ends with U+002F (/);
+            specifier_key.ends_with("/"sv) &&
+            // specifierKey is a code unit prefix of normalizedSpecifier; and
+            Infra::is_code_unit_prefix(specifier_key, normalized_specifier) &&
+            // either asURL is null, or asURL is special,
+            (!as_url.has_value() || as_url->is_special())) {
+            // then:
```

Comment:
What I did for fetch is put the then comment before the closing paren, gets formatted a bit nicer and helps with readability

```cpp
if (
    ...
    // - either asURL is null, or asURL is special,
    (!as_url.has_value() || as_url->is_special())
    // then:
) {
    // 1. If resolutionResult is null ...
```